cudax/include/cuda/experimental/__device/arch_traits.cuh

```diff
@@ -0,0 +1,401 @@
+  __traits.redux_intrinisic   = true;
+  __traits.elect_intrinsic    = false;
+  __traits.cp_async_supported = true;
+  __traits.tma_supported      = false;
+
+  return __traits;
+}();
+
+inline constexpr arch_traits_t sm_900_traits = []() constexpr {
+  arch_traits_t __traits{};
+  __traits.compute_capability_major             = 9;
+  __traits.compute_capability_minor             = 0;
+  __traits.compute_capability                   = 90;
+  __traits.max_shared_memory_per_multiprocessor = 228 * 1024;
+  __traits.max_blocks_per_multiprocessor        = 32;
+  __traits.max_warps_per_multiprocessor         = 64;
+  __traits.max_threads_per_multiprocessor =
+    __traits.max_warps_per_multiprocessor * detail::arch_common_traits::warp_size;
+  __traits.reserved_shared_memory_per_block = 1024;
+  __traits.max_shared_memory_per_block_optin =
+    __traits.max_shared_memory_per_multiprocessor - __traits.reserved_shared_memory_per_block;
+
+  __traits.cluster_supported  = true;
+  __traits.redux_intrinisic   = true;
+  __traits.elect_intrinsic    = true;
+  __traits.cp_async_supported = true;
+  __traits.tma_supported      = true;
+
+  return __traits;
+}();
+
+// TODO Should this be provided outside detail? Just using arch template seems better
+template <unsigned int __SmVersion>
+_CCCL_HOST_DEVICE inline constexpr arch_traits_t arch_traits()
+{
+  if constexpr (__SmVersion == 60)
+  {
+    return detail::sm_600_traits;
+  }
+  else if constexpr (__SmVersion == 70)
+  {
+    return detail::sm_700_traits;
+  }
+  else if constexpr (__SmVersion == 75)
+  {
+    return detail::sm_750_traits;
+  }
+  else if constexpr (__SmVersion == 80)
+  {
+    return detail::sm_800_traits;
+  }
+  else if constexpr (__SmVersion == 86)
+  {
+    return detail::sm_860_traits;
+  }
+  else
+  {
+    static_assert(__SmVersion == 90, "Unknown architecture");
+    return detail::sm_900_traits;
+  }
+}
+
+} // namespace detail
+
+//! @brief Retrieve architecture traits of the specified architecture
+//!
+//! @param __sm_version Compute capability in 10 * major + minor format for which the architecture traits are requested
+//!
+//! @throws cuda_error if the requested architecture is unknown
+_CCCL_HOST_DEVICE inline constexpr arch_traits_t arch_traits(unsigned int __sm_version)
+{
+  switch (__sm_version)
+  {
+    case 60:
+      return detail::sm_600_traits;
+    case 70:
+      return detail::sm_700_traits;
+    case 75:
+      return detail::sm_750_traits;
+    case 80:
+      return detail::sm_800_traits;
+    case 86:
+      return detail::sm_860_traits;
+    case 90:
+      return detail::sm_900_traits;
+    default:
+      __throw_cuda_error(cudaErrorInvalidValue, "Traits requested for an unknown architecture");
+      break;
+  }
+}
+
+//! @brief Type representing a CUDA device architecture. It provides traits from arch_traits_t in form of static members
+template <unsigned int __SmVersion>
+struct arch : public detail::arch_common_traits
```

Comment:
Question: Why do we need an additional template? For every `arch<80>::property`, I can also call `arch_traits(80).property`.

Reply:
The idea is to provide a type that you can use to template your kernel for tuning etc. With just a struct its not possible in c++17, it could be an int template argument that is passed into arch_traits, but it seems nicer to have it strongly typed

Reply:
Ah, I see. You want to pass:
```c++
kernel<arch<80>>();
```
Passing as NTTP would require C++20 I think:
```c++
kernel<arch_traits(80)>();
```
However, looking towards the future, we can consider how users are going to write code over the next decade or two. But yeah, having the template as a stop-gap solution is a valid reason.

Also notice, that we could not use it in CUB that way, because we need the exact same template instantiation for all arches we are compiling for. So in CUB, we must instantiate the arch template inside the kernel. Also, since all information is statically availabele, I can just `arch_traits(__CUDA_ARCH__ / 10)` in all the places I need, without any (runtime) overhead.

Reply:
In general I like the type more here. One of the things I would like to do in the future is to allow per-architecture kernel configuration.
Right now with `cudax::launch` you pass in the kernel configuration struct, but we could accept some form of map from architecture to configuration. Having `arch_traits_t` as the key seems less intuitive than instances of `arch<sm>` type.
Also, lets say I want to go a step further and provide kernel config per specific SKU, for example tuned for specific number of SMs. The map where keys are separete types would allow me to do it more easily.

So in general I think there are enough small reasons to provide the `arch` template in addition to `arch_traits_t`, even if it seems a bit redundant.

About the CUB usage, we will see, maybe we will be able to convince the compiler to do something that would allow CUB to drop the chained policy 😉